src/librustc_middle/mir/interpret/queries.rs

```diff
@@ -74,4 +74,13 @@ impl<'tcx> TyCtxt<'tcx> {
             self.const_eval_validated(inputs)
         }
     }
+
+    /// Evaluate a static, returning the allocation of the initializer's memory.
+    pub fn const_eval_static(self, def_id: DefId) -> Result<&'tcx Allocation, ErrorHandled> {
```

Comment:
yes and no. During codegen we'll want to use integer constants as llvm integer constants. If we used a scheme like this for constants, too, we'd have to repeat the conversion to a `ConstValue` (or some similar structure).

Reply:
we could read integers from the allocation directly though, so it may not be too bad... I'll experiment

Reply:
We can do that in a separate PR though. Let's discuss in https://github.com/rust-lang/rust/issues/72396.

Reply:
Since we still have (and maybe forever will have) `ty::Const` as an optimization for literals, even in `mir::Const`, we need to support this anyway, so I think for now we should just do this for statics where it's unambiguous.

Reply:
> ty::Const as an optimization for literals, even in mir::Const

I do not understand this statement. Are you saying `mir::Const` will not just be an `AllocId`? I think that is the wrong approach, I'd rather make the caller that has these weird special needs (i.e., codegen for consts) deal with "I want raw representation except sometimes", than making *everyone* deal with it.

Reply:
`mir::Const` is a hypothetical thing that we haven't specified. We don't need it at all if I'm understanding your proposal correctly, as it would just be a `&'tcx Allocation` (or `AllocId`). What we do need is for the `mir::Body` to contain both `ty::Const` (for const generics, unevaluated and literals) as well as `&'tcx Allocation` for successfully evaluated constants.

Reply:
Please let's stop this off-topic discussion here.

Reply:
It is getting a bit frustrating that you keep ignoring my requests to move the discussion, not even explaining why you think that the discussion should not be moved.

Reply:
Sorry, I completely overlooked the first one and the second one was edited in late so that one I blame on github.